sql/src/main/java/io/crate/operation/projectors/SortingProjector.java

```
@@ -0,0 +1,140 @@
+                     int offset) {
+        Preconditions.checkArgument(offset >= 0, "invalid offset %s", offset);
+        this.numOutputs = numOutputs;
+        this.inputs = inputs;
+        this.collectExpressions = collectExpressions;
+        this.ordering = ordering;
+        this.offset = offset;
+    }
+
+    @Override
+    public boolean setNextRow(Row row) {
+        for (CollectExpression<Row, ?> collectExpression : collectExpressions) {
+            collectExpression.setNextRow(row);
+        }
+        Object[] newRow = new Object[inputs.size()];
+        int i = 0;
+        for (Input<?> input : inputs) {
+            newRow[i++] = input.value();
+        }
+        rows.add(newRow);
+        return true;
+    }
+
+    @Override
+    public void finish() {
+        if (rows.size() == 0) {
+            downstream.finish();
+            return;
+        }
+        // sort, we must reverse the order (back to original one) because order was reserved for used on queues
+        Collections.sort(rows, Collections.reverseOrder(ordering));
+
+        // process offset
+        if (offset != 0) {
+            rows = rows.subList(offset, rows.size());
+        }
+
+        // emit
+        rowEmitter = createRowEmitter();
+        rowEmitter.run();
+    }
+
+    private IterableRowEmitter createRowEmitter() {
+        return new IterableRowEmitter(downstream, new CollectionBucket(rows, numOutputs));
+    }
+
+    @Override
+    public void kill(Throwable throwable) {
+        IterableRowEmitter emitter = rowEmitter;
+        if (emitter == null) {
+            downstream.kill(throwable);
+        } else {
+            emitter.kill(throwable);
+        }
+    }
+
+    @Override
+    public void fail(Throwable t) {
+        downstream.fail(t);
+    }
+
+    @Override
+    public void repeat() {
+        // just emit all again, sorting and offset processing is expected to be done already
+        rowEmitter = createRowEmitter();
```

I think repeat should throw an error here...  Rows are only send to the downstream via the RowEmitter which becomes the new upstream - so the RowEmitter should receive the repeat call, not the projector directly